### PR TITLE
fix Issue 19463 - DIP1008 - _d_newclass is called instead of _d_newTh…

### PR DESCRIPTION
**Pull request by WalterBright**
…rowable

Well, this is embarrassing. I must have never checked that code in. The trouble is, I'm not sure how to test it, other than manually checking the output which I've done.

### CODE REVIEW

**Comment by thewilsonator** [comment]
A simple shell test of `! ( dmd file.d -o file.o && nm file.o | grep _d_new_class)` should do the trick.

**Comment by Geod24** [comment]
> Well, this is embarrassing. I must have never checked that code in. The trouble is, I'm not sure how to test it, other than manually checking the output which I've done.

Two options I can think of:
- In a test, define `_d_newClass` to be `assert(0)` so calling it would crash the binary
- Check the resulting binary's symbol table

The first option has the advantage of being rather simple and platform-independent

**Comment by WalterBright** [comment]
Unfortunately, even in a trivial program _d_newclass is being called from elsewhere, and that idea fails. So it's the script plan.

**Comment by thewilsonator** [comment]
>program

Thats why I suggested an object file.

**Comment by WalterBright** [comment]
@thewilsonator I added a script, can you please have a look at it? Not very experienced with scripts.

**Comment by thewilsonator** [comment]
I hate script tests too, but that looks fine to me. We'll have to see what the autotester thinks.

**Comment by WalterBright** [comment]
@atilaneves can you please look into the buildkite/dmd failure, as it is kaliedoscope

**Comment by thewilsonator** [comment]
Seems to be a problem with exception.toString

**Comment by thewilsonator** [comment]
e.g. 
```
tests/ut/wrap/traits.d:61 - Expected: "Unsupported function type double(int) for foo"
tests/ut/wrap/traits.d:61 -      Got: ""
```

**Comment by WalterBright** [comment]
This PR also appears to resolve https://issues.dlang.org/show_bug.cgi?id=19317

**Comment by jacob-carlborg** [comment]
Isn't it better to check that `_d_newThrowable` is called? Implement that function, in the implementation change a TLS value and assert that is change correctly.

> Unfortunately, even in a trivial program _d_newclass is being called from elsewhere, and that idea fails. 

Perhaps check which ClassInfo is passed an assert only when it matches the ClassInfo of the thrown class.

**Comment by atilaneves** [comment]
I pushed and tagged a new excel-d, should work now.

**Comment by WalterBright** [comment]
Added test case from https://issues.dlang.org/show_bug.cgi?id=19317

**Comment by WalterBright** [comment]
Thanks, @atilaneves !

**Comment by WalterBright** [comment]
@atilaneves sadly, it's still failing kaleidoscope

**Comment by atilaneves** [comment]
@WalterBright It's failing on the old tag (0.4.1), I pushed the new tag (0.4.2) to the wrong remote by mistake. Should work now.

**Comment by WalterBright** [comment]
Thank you, @atilaneves ! Now, if only @thewilsonator or @wilzbach would approve this!

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
> This PR also appears to resolve https://issues.dlang.org/show_bug.cgi?id=19317

Please adjust the commit message to reflect that so the dlang bot picks up on it.

**Comment by WalterBright** [comment]
I resolved 19317 as a duplicate of 19463, and the test case for it was added to this PR.

**Comment by thewilsonator** [comment]
OK.

File: test/compilable/test19463.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+
+$DMD -c -dip1008 -m${MODEL} -of${OUTPUT_BASE}${OBJ} -I${EXTRA_FILES} ${EXTRA_FILES}/${TEST_NAME}.d
+!(nm ${OUTPUT_BASE}${OBJ} | grep _d_newclass)
```
Comment (jacob-carlborg) [inline test/compilable/test19463.sh]: @WalterBright As far as I understand this verifies that `_d_newclass` is **not** called, not that `_d_newThrowable` **is** called. I recommend implementing `_d_newThrowable` instead and verify it's getting called. This way a shell script is not required.

**Comment by rainers** [comment]
> Well, this is embarrassing.

Still embarrassing that this now always calls _d_newThrowable, irrespective of whether -dip1008 is specified or not.

**Pull request merged**